kernel/src/utilities/capability_ptr.rs

```diff
@@ -0,0 +1,161 @@
+}
+
+impl Default for CapabilityPtr {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null(),
+        }
+    }
+}
+
+/// Permission sets a [`CapabilityPtr`] may grant.
+/// These may not be enforced or exist on a given platform.
+#[derive(Copy, Clone, PartialEq)]
+pub enum CapabilityPtrPermissions {
+    None,
+    Read,
+    Write,
+    ReadWrite,
+    Execute,
+}
+
+impl From<CapabilityPtr> for usize {
+    /// Returns the address of the [`CapabilityPtr`].
+    /// Provenance note: may not expose provenance.
+    #[inline]
+    fn from(from: CapabilityPtr) -> Self {
+        from.ptr as usize
+    }
+}
+
+impl From<usize> for CapabilityPtr {
+    /// Constructs a [`CapabilityPtr`] with a given address and no authority
+    ///
+    /// Provenance note: may have null provenance.
+    #[inline]
+    fn from(from: usize) -> Self {
+        Self {
+            ptr: from as *const (),
+        }
+    }
+}
+
+impl UpperHex for CapabilityPtr {
+    /// Format the capability as an uppercase hex string.
+    /// Will print at least the address, and any platform specific metadata if it exists.
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        UpperHex::fmt(&(self.ptr as usize), f)
+    }
+}
+
+impl LowerHex for CapabilityPtr {
+    /// Format the capability as a lowercase hex string.
+    /// Will print at least the address, and any platform specific metadata if it exists.
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        LowerHex::fmt(&(self.ptr as usize), f)
+    }
+}
+
+impl AddAssign<usize> for CapabilityPtr {
+    /// Increments the address of a [`CapabilityPtr`]
+    #[inline]
+    fn add_assign(&mut self, rhs: usize) {
+        self.ptr = (self.ptr as *const u8).wrapping_add(rhs) as *const ();
+    }
+}
+
+impl CapabilityPtr {
+    /// Returns the pointer component of a [`CapabilityPtr`] but without any of the authority.
+    pub fn as_ptr<T>(&self) -> *const T {
+        self.ptr as *const T
+    }
+
+    /// Construct a [`CapabilityPtr`] from a raw pointer, with authority ranging over
+    /// [`base`, `base + length`) and permissions `perms`.
+    ///
+    /// Provenance note: may derive from a pointer other than the input to provide something with
+    /// valid provenance to justify the other arguments.
+    ///
+    /// ## Safety
+    ///
+    /// Constructing a [`CapabilityPtr`] with metadata may convey authority to
+    /// dereference this pointer, such as in userspace. When these pointers
+    /// serve as the only memory isolation primitive in the system, this method
+    /// can thus break Tock's isolation model. As semi-trusted kernel code can
+    /// name this type and method, it is thus marked as `unsafe`.
+    ///
+    // TODO: Once Tock supports hardware that uses the [`CapabilityPtr`]'s
+    // metdata to convey authority, this comment should incorporate the exact
+    // safety conditions of this function.
```

Comment:
This will happen soon, correct? We already have three calls to this function with no safety justification.

It's unclear to me what the safety conditions will look like (i.e. how we will prevent capsules from breaking isolation by copying `CapabilityPtr`s into userspace memory).

Reply:
Technically, there are no safety requirements from rust's perspective (I would have voted "no" to making this interface unsafe) for the use of this type within the kernel. This type does not implement Deref, and is no more unsafe than the MPU configuration interfaces, which are not unsafe at present.

If we did decide that any of these might eventually be passed to the user, and there was a library invariant they should never be allowed access to kernel memory, I suppose it might say "This should never create an alias with any objects accessible by the kernel, apart from those accessed through the grant interface".

Reply:
> It's unclear to me what the safety conditions will look like (i.e. how we will prevent capsules from breaking isolation by copying `CapabilityPtr`s into userspace memory).

Part of the reason of why I added this comment is exactly because these and other details are unclear to me. There will be safety considerations, and this is a public API, and as such this is to serve as a deterrent for anyone to rely on any stable safety conditions at this point.

> Technically, there are no safety requirements from rust's perspective (I would have voted "no" to making this interface unsafe) for the use of this type within the kernel.

In a system where capabilities serve as the only memory isolation mechanism, we must ensure that any untrusted code we're invoking (such as a process) cannot violate any of Rust's safety invariants. If arbitrary capabilities can flow across this boundary on a type-level (such as ones pointing into kernel memory, originating from capsules) either the boundary would need to be unsafe, or creating these capabilities must be.

This argument relies on the premise that there might be an API that would capsules to smuggle capabilities into userspace. Will this API never exist?

> This type does not implement Deref, and is no more unsafe than the MPU configuration interfaces, which are not unsafe at present.

I think this is the crux: I strongly believe that the MPU configuration interfaces ought to be unsafe, following this same reasoning. We have some much weaker argument for why it might be okay for them not to be unsafe: capsules shouldn't be able to get a hold of a reference the chip's MPU type, and as such they wouldn't be able to use those "safe" interfaces. I disagree with this line of reasoning, and there likely is a way for capsules to get a reference like this. However, that is a different discussion, orthogonal to this PR.

Reply:
I definitely agree with your premise that we have a library invariant whenever MPUs / CHERI are used that we do not allow the user access to kernel, and this very much can become language UB if the user trashes the kernel.

My objections were mostly that CHERI suddenly seems special as other MPU accesses are not considered unsafe, and on the practiically of marking the interface unsafe.

If we do mark the interface unsafe, we should surely justify at the call sites that we have enforced insolation, or else propagate unsafety requirements upwards. My suspicion is that isolation probably can't be locally argued and is likely a property of broad kernel correctness. Possibly, the scope of the entire core kernel, so we would either just have to admit defeat and say every function is unsafe, or instead lie at call sites and say we we can be certain we are not breaking safety here, even though the scope is far too broad to ever do so in full spirit.

TL;DR - this can be unsafe, but then a whole lot more should be too.

Reply:
I think I agree to the above in spirit, although I think we will be able to end up precisely drawing the line around unsafe interfaces, and hence don't agree with the generalization in the third paragraph.

In any case, let's wait on the follow-up PR that will introduce the reality of a real capability-hardware platform where these abstract concerns will manifest into tangible isolation properties / violations. Then we can revisit these call-sites and add appropriate documentation.